examples/pagerank.cpp

```diff
@@ -0,0 +1,164 @@
+        this->key = key;
+    }
+
+    virtual void add_neighbors(KeyT nbKey) {
+        this->nbs.push_back(nbKey);
+    }
+
+    virtual std::vector<KeyT> & get_neighbors() {
+        return this->nbs;
+    }
+
+    static uint64_t partition(const KeyT & key) {
+        return key;
+    }
+
+    friend husky::BinStream & operator >> (husky::BinStream & stream, husky::Graph::PRVertex & v) {
+        stream >> v.pr >> v.nbs >> v.key;
+        return stream;
+    }
+
+    friend husky::BinStream & operator << (husky::BinStream & stream, husky::Graph::PRVertex v) {
+        stream << v.pr << v.nbs << v.key;
+        return stream;
+    }
+
+protected:
+    std::vector<KeyT> nbs;
+};
+
+std::function<void(PRVertex &)> exec_pagerank(husky::ObjList<PRVertex> & obj_list, \
+ husky::PushCombinedChannel<float, PRVertex, SumCombiner<float>> & ch) {
+    std::function<void(PRVertex &)> exec_lambda = [&](PRVertex & v) {
+        float sum = 0;
+        sum += ch.get(v);
+        v.pr = 0.15 + 0.85 * sum;
+
+        if (v.get_neighbors().size() == 0) return;
+
+        float pr_contrib = v.pr / v.get_neighbors().size();
+        for (auto & nb : v.get_neighbors()) {
+            ch.push(pr_contrib, nb);
+        }
+    };
+    return exec_lambda;
+}
+
+}  // namespace Graph
+
+}  // namespace husky
+
+// using namespace husky;
+
+void pagerank() {
+    auto start = std::chrono::steady_clock::now();
+    io::HDFSLineInputFormat infmt;
+    infmt.set_input(husky::Context::get_param("input"));
+    husky::ObjList<husky::Graph::PRVertex> pr_list;
+    auto & ch = husky::ChannelFactory::create_push_combined_channel<float, husky::SumCombiner<float>>(pr_list, pr_list);
+    auto parse_pr = [&](boost::string_ref & chunk) {
+        if (chunk.size() == 0) return;
+        boost::char_separator<char> sep(" \t");
+        boost::tokenizer<boost::char_separator<char>> tok(chunk, sep);
+        husky::Graph::PRVertex obj;
+        int i = 0;
+        for (auto & v : tok) {
+            i++;
+            if (i == 1)
+                obj.key_init(std::stoi(v));
+            else if (i == 2);
+            else
+                obj.add_neighbors(std::stoi(v));
+        }
+        pr_list.add_object(obj);
+    };
+
+    husky::load(infmt, parse_pr);
+    husky::globalize(pr_list);
+
+    husky::list_execute(pr_list, [&](husky::Graph::PRVertex & v) {
+        // send dummy page rank values
+        for (auto & nb : v.get_neighbors()) {
+            ch.push(0, nb);
+        }
+    });
+
+    if (husky::Context::get_global_tid() == 0) {
+        base::log_msg("Load time: " + std::to_string(std::chrono::duration_cast<std::chrono::duration<float>> \
+         (std::chrono::steady_clock::now()-start).count()));
+    }
+
+    husky::list_execute(pr_list, husky::Graph::exec_pagerank(pr_list, ch), \
+     std::stoi(husky::Context::get_param("num_iters")));
+
+    husky::list_execute(pr_list, [](husky::Graph::PRVertex & v) {base::log_msg("global threadID: " + \
```

Comment:
I have modified the codes. In the modified version, I let all the worker send their message in the form of a string containing the vertexID and the vertex pagerank value to one worker and let that worker sort and output the top k page rank values. Which one do you think is better because the latter on will also introduce a lot of traffic if the graph is very big. Can you review my lattest commit?